### PR TITLE
[async] Support SFG-level DSE for scalar SNodes

### PR DESCRIPTION
**Pull request by k-ye**
Sorry for this large change. It shouldn't be as complicated as the size implies.. The approach is basically to find out those **scalar** SNodes whose values are not read by the successor tasks. We then pass in these SNodes to the CFG pass, and instruct the live variable analysis **not to** add them into the final node's `live_gen`.

There are a few special handlings for scalar SNodes here:

1. Upon seeing a `GlobalStoreStmt` when constructing the task meta, we add the SNode to the input (reader) states, iff. it is *not* a scalar. https://github.com/taichi-dev/taichi/blob/3aeff7c01f1dff6def9602b7f4a0005fc54b4c22/taichi/program/async_utils.cpp#L127-L130
1. In `SFG::optimize_dead_store()`, if the state forms a flow edge, we also check if it's a scalar and if the successor task actually reads it. https://github.com/taichi-dev/taichi/blob/3aeff7c01f1dff6def9602b7f4a0005fc54b4c22/taichi/program/state_flow_graph.cpp#L800-L801
1. atomic demotion for global values are only enabled for `serial` tasks. I also made them handle scalars as well. https://github.com/taichi-dev/taichi/blob/3aeff7c01f1dff6def9602b7f4a0005fc54b4c22/taichi/ir/control_flow_graph.cpp#L403

Finally, the optimization result is cached, much like what @xumingkuan  has done in #1889.

Related issue = #742

[[Click here for the format server]](http://kun.csail.mit.edu:31415/)

----

### CODE REVIEW

**Review by yuanming-hu — APPROVED** [review]
Thank you! I would leave the `control_flow_graph` part to @xumingkuan but everything else LGTM!

File: tests/python/test_sfg.py [hunk 1]
```diff
@@ -58,3 +59,48 @@ def serial_z():
         else:
             assert ys[i] == i + 2
             assert xs[i] == 0
+
+
+@ti.test(require=ti.extension.async_mode, async_mode=True)
+def test_sfg_dead_store_elimination():
+    ti.init(arch=ti.cpu,
+            async_mode=True,
+            async_opt_intermediate_file="simple_diff")
```
[the remark this answers is not in the document]
Comment (k-ye) [reply tests/python/test_sfg.py:68]: Ah yeah, sorry i forgot to remove these debugging stuff..

File: tests/python/test_sfg.py [hunk 2]
```diff
@@ -58,3 +59,48 @@ def serial_z():
...
+            unused[None] += x[i]
+
+    xnp = np.arange(n, dtype=np.float32)
+    x.from_numpy(xnp)
+    ti.sync()
+
+    stats = ti.get_kernel_stats()
+    stats.clear()
+
+    for _ in range(5):
```
[the remark this answers is not in the document]
Comment (k-ye) [reply tests/python/test_sfg.py:91]: Actually the DSE works pretty deterministically :) I just wasn't sure if it was a good idea to be very specific about how many DSE has been done..

**Review by xumingkuan — APPROVED** [review]
The `control_flow_graph` part LGTM!

**Pull request merged**